packages/provider/index.js

```diff
@@ -26,19 +29,13 @@ module.exports = {
         { keepAlive: false }
       );
     }
-
-    return this.wrap(provider, options);
+    return provider;
   },
 
-  test_connection: function(provider, callback) {
-    var web3 = new Web3Shim({ provider });
-    var fail = new Error(
-      "Could not connect to your RPC client. Please check your RPC configuration."
-    );
-
-    web3.eth
-      .getCoinbase()
-      .then(coinbase => callback(null, coinbase))
-      .catch(() => callback(fail, null));
+  testConnection: async function(options) {
+    const provider = this.getProvider(options);
```

Comment:
Why do we need this `getProvider` call?

Reply:
On second thought, I think we should remove it. It couples us more closely to Web3 & EVM-like networks.

Reply:
Why do you think it couples us? This method is our adapter for getting the provider. Sometimes people provide it as a function, sometimes directly. This handles getting whatever provider they give us.

Reply:
Most of the time users don't specify a provider in their config. With this code, they'll be hitting the `else` clause in `getProvider`.

Thinking about this some more, why not `try/catch` around the `getBlockNumber` to test the provider?

Actually, we already de facto do a number of provider checks in [`Environment.detect`](https://github.com/trufflesuite/truffle/blob/develop/packages/environment/environment.js#L11-L24). I agree the error messaging should/could be better. It looks like there could be a number of reasons why the provider fails.

Reply:
This PR does not change the logic for creating a provider but just organizes it a bit differently. Are you saying we should change the fundamental logic with how the provider is currently being created in the `create` method?

Are you also saying that it would be better to change the signature of the `testConnection` method? Currently it returns `true` if it is successful and throws an error if not. Perhaps it should be name `connectionValid` and return `true` or `false depending on whether the provider seems valid or not.

I mean, currently there are many cases where the process just dies without giving the user feedback. This was the primary reason I wanted to put in this PR.

Reply:
I think the `create` logic is fine for now since we're able to re-route what the provider does via the `Web3Shim`/`InterfaceAdapter`.

I guess not only do I think the implementation of `testConnection` into the `migrate` flow is redundant (it won't get hit until a number of web3 calls have already been made), but having...

wait nm, for non-evm networks we're likely not going to do anything with the provider (for now).

So basically, integrating `testConnection` into the `migrate` flow seems redundant. Do we have reproduction repo's or have we been able to isolate and pinpoint where exactly users are having their process' die?

Reply:
That's a good point that the provider is used before this call so maybe this call to `testConnection` method is kind of nonsensical. I haven't reproduced this in a while and if I remember correctly it was some call to `getId` which might be in `Environment.detect`. 

Let me try and see if I can reproduce that problem later today to see how it might be handled better. Currently if I feed it a bogus provider it seems to error properly with a nice error message. If I can reproduce this problem of it dying silently then maybe that will be a better time to implement this call to test if the provider is good.

Reply:
So if you give it a random url ("www.google.com" for example) it will timeout and give no message. We need to figure out a way to test this. It looks like `getId` will not cut it for `testConnection`.